src/libraries/Common/src/System/TimeProvider.cs

```diff
@@ -10,115 +10,100 @@ namespace System
-        /// a time zone based on <paramref name="timeZone"/>, a high-performance time stamp based on <see cref="Stopwatch"/>,
-        /// and a timer based on <see cref="Timer"/>.
-        /// </summary>
-        /// <param name="timeZone">The time zone to use in getting the local time using <see cref="LocalNow"/>. </param>
-        /// <returns>A new instance of <see cref="TimeProvider"/>. </returns>
-        /// <exception cref="ArgumentNullException"><paramref name="timeZone"/> is null.</exception>
-        public static TimeProvider FromLocalTimeZone(TimeZoneInfo timeZone)
-        {
-#if SYSTEM_PRIVATE_CORELIB
-            ArgumentNullException.ThrowIfNull(timeZone);
-#else
-            if (timeZone is null)
-            {
-                throw new ArgumentNullException(nameof(timeZone));
-            }
-#endif // SYSTEM_PRIVATE_CORELIB
-
-            return new SystemTimeProvider(timeZone);
-        }
+        /// <remarks>
+        /// The default implementation returns <see cref="Stopwatch.Frequency"/>.
+        /// </remarks>
+        public virtual long TimestampFrequency { get => Stopwatch.Frequency; }
 
         /// <summary>
         /// Gets the current high-frequency value designed to measure small time intervals with high accuracy in the timer mechanism.
         /// </summary>
         /// <returns>A long integer representing the high-frequency counter value of the underlying timer mechanism. </returns>
-        public abstract long GetTimestamp();
+        /// <remarks>
+        /// The default implementation returns <see cref="Stopwatch.GetTimestamp"/>.
+        /// </remarks>
+        public virtual long GetTimestamp() => Stopwatch.GetTimestamp();
 
         /// <summary>
         /// Gets the elapsed time between two timestamps retrieved using <see cref="GetTimestamp"/>.
         /// </summary>
         /// <param name="startingTimestamp">The timestamp marking the beginning of the time period.</param>
         /// <param name="endingTimestamp">The timestamp marking the end of the time period.</param>
         /// <returns>A <see cref="TimeSpan"/> for the elapsed time between the starting and ending timestamps.</returns>
-        public TimeSpan GetElapsedTime(long startingTimestamp, long endingTimestamp) =>
-            new TimeSpan((long)((endingTimestamp - startingTimestamp) * _timeToTicksRatio));
+        public TimeSpan GetElapsedTime(long startingTimestamp, long endingTimestamp)
+        {
+            long timestampFrequency = TimestampFrequency;
+            if (timestampFrequency <= 0)
+            {
+#if SYSTEM_PRIVATE_CORELIB
+                ArgumentOutOfRangeException.ThrowIfNegativeOrZero(TimestampFrequency);
+#else
+                string propertyName = nameof(TimestampFrequency);
+                throw new ArgumentOutOfRangeException(propertyName, timestampFrequency, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonNegativeNonZero, nameof(TimestampFrequency)));
```

Comment:
```suggestion
                throw new ArgumentOutOfRangeException(nameof(TimestampFrequency), timestampFrequency, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonNegativeNonZero, nameof(TimestampFrequency)));
```

Reply:
I had it as you suggested but it is producing the following error:

```diff 
- error CA2208: Method GetElapsedTime passes 'TimestampFrequency' as the paramName argument to a ArgumentOutOfRangeException constructor. Replace this argument with one of the method's parameter names. Note that the provided parameter name should have the exact casing as declared on the method. [C:\oss\runtime\src\libraries\Microsoft.Bcl.TimeProvider\src\Microsoft.Bcl.TimeProvider.csproj::TargetFramework=net462]
```

That is why I worked around it by using `string propertyName = nameof(TimestampFrequency);`

Reply:
I hadn't looked to see what this was for. I think ArgOutOfRangeException is the wrong type here.... it's not an argument that's wrong. The analyzer is right that something is amiss.

Reply:
InvalidOperationException will be more reasonable here?

Reply:
Yeah, I think that's better.

Reply:
Done!